File: 322_Coin_Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0)
            return 0;
        vector <int> f(amount + 1 , INT_MAX);
        f[0] = 0;
        for (size_t i = 1; i != f.size(); ++i) {
            for (size_t j = 0; j != coins.size(); ++j) {
                if (i < coins[j] || f[i-coins[j]] == INT_MAX)
                    continue;
                f[i] = min(f[i-coins[j]] + 1, f[i]);
            }
        }
        if (f[amount] == INT_MAX)
            return -1;
        return f[amount];
    }
};

File: 322_Coin_Change.py
import numpy as np


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        dp, coins = np.zeros(amount + 1).astype(int) - 1, np.asarray(coins)
        dp[0] = 0
        for i in range(1, len(dp)):
            relev_cases = dp[i - coins[coins <= i]]
            valid_cases = relev_cases[relev_cases != -1]
            if len(valid_cases) > 0:
                dp[i] = min(valid_cases) + 1
        return dp[-1] if dp[-1] > 0 else -1
